File: Day1-9/9.py
def stream_score(stream):
    score, chars = 0, 0
    idx = 0
    stack = [0]
    while idx < len(stream):
        if stream[idx] == '{':
            score += stack[-1] + 1
            stack.append(len(stack))
            idx += 1
        elif stream[idx] == '}':
            stack.pop()
            idx += 1
        elif stream[idx] == '<':
            g_len, g_chars = garbage_length(stream[idx:])
            idx += g_len
            chars += g_chars
        else:
            idx += 1
    return score, chars


def garbage_length(stream):
    '''
    Given a stream of format "<..."
    calculate the length of the garbage group
    '''
    chars = 0
    idx = 1
    while idx < len(stream):
        if stream[idx] == '>':
            break
        elif stream[idx] == '!':
            idx += 2
        else:
            idx += 1
            chars += 1
    return idx + 1, chars


if __name__ == '__main__':
    # assert garbage_length('<random characters>') == len('<random characters>')
    # assert garbage_length('<<<<>') == len('<<<<>')
    # assert garbage_length('<{!>}>') == len('<{!>}>')
    # assert garbage_length('<!!>') == len('<!!>')
    # assert garbage_length('<!!!>>') == len('<!!!>>')
    # assert garbage_length('<{o"i!a,<{i<a>') == len('<{o"i!a,<{i<a>')

    # assert stream_score('{}') == 1
    # assert stream_score('{{{}}}') == 6
    # assert stream_score('{{},{}}') == 5
    # assert stream_score('{{<a!>},{<a!>},{<a!>},{<ab>}}') == 3

    with open('9.txt') as f:
        stream = f.read().strip()
    score, chars = stream_score(stream)
    print("Part 1: {}".format(score))
    print("Part 2: {}".format(chars))

File: Day1-9/9.go
package main

import (
	"fmt"
	"io/ioutil"
)

func parseStream(stream []byte) (streamScore int, garbageChars int) {
	streamScore = 0
	garbageChars = 0
	currLevel := 0

	for idx := 0; idx < len(stream); {
		switch stream[idx] {
		case '{':
			streamScore += currLevel + 1
			currLevel++
			idx++
		case '}':
			currLevel--
			idx++
		case '<':
			gLen, gChars := parseGarbage(stream[idx:])
			idx += gLen
			garbageChars += gChars
		default:
			idx++
		}
	}
	return
}

func parseGarbage(stream []byte) (garbageLen int, numChars int) {
	numChars = 0

parseLoop:
	for garbageLen = 1; garbageLen < len(stream); {
		switch stream[garbageLen] {
		case '>':
			break parseLoop
		case '!':
			garbageLen += 2
		default:
			garbageLen += 1
			numChars++
		}
	}
	return
}

func main() {
	dat, _ := ioutil.ReadFile("./9.txt")
	streamScore, garbageChars := parseStream(dat)
	fmt.Printf("Part 1: %d\n", streamScore)
	fmt.Printf("Part 2: %d\n", garbageChars)
}
